messaging/src/main/java/org/axonframework/commandhandling/distributed/AbstractRoutingStrategy.java

```diff
@@ -17,62 +17,99 @@
+     *
+     * @param builder the {@link Builder} used to instantiate a {@link AbstractRoutingStrategy} instance
+     */
+    protected AbstractRoutingStrategy(Builder<B> builder) {
+        builder.validate();
+        this.fallbackRoutingStrategy = builder.fallbackRoutingStrategy;
+    }
 
     /**
-     * Initializes the strategy using given {@code unresolvedRoutingKeyPolicy} prescribing what happens when a
-     * routing key cannot be resolved.
+     * Initializes the strategy using given {@link UnresolvedRoutingKeyPolicy} prescribing the fallback approach when
+     * this implementation cannot resolve a routing key.
      *
-     * @param unresolvedRoutingKeyPolicy The policy for unresolved routing keys.
+     * @param fallbackRoutingStrategy the fallback routing to use whenever this {@link RoutingStrategy} doesn't succeed
+     * @deprecated in favor of the {@link #AbstractRoutingStrategy(Builder)}
      */
-    public AbstractRoutingStrategy(UnresolvedRoutingKeyPolicy unresolvedRoutingKeyPolicy) {
-        Assert.notNull(unresolvedRoutingKeyPolicy, () -> "unresolvedRoutingKeyPolicy may not be null");
-        this.unresolvedRoutingKeyPolicy = unresolvedRoutingKeyPolicy;
+    @Deprecated
+    public AbstractRoutingStrategy(UnresolvedRoutingKeyPolicy fallbackRoutingStrategy) {
+        assertNonNull(fallbackRoutingStrategy, "Fallback RoutingStrategy may not be null");
+        this.fallbackRoutingStrategy = fallbackRoutingStrategy;
     }
 
     @Override
     public String getRoutingKey(CommandMessage<?> command) {
         String routingKey = doResolveRoutingKey(command);
         if (routingKey == null) {
-            switch (unresolvedRoutingKeyPolicy) {
-                case ERROR:
-                    throw new CommandDispatchException(format("The command [%s] does not contain a routing key.",
-                                                              command.getCommandName()));
-                case RANDOM_KEY:
-                    return Long.toHexString(counter.getAndIncrement());
-                case STATIC_KEY:
-                    return STATIC_ROUTING_KEY;
-                default:
-                    throw new IllegalStateException("The configured UnresolvedRoutingPolicy of "
-                                                            + unresolvedRoutingKeyPolicy.name() + " is not supported.");
-            }
+            routingKey = fallbackRoutingStrategy.getRoutingKey(command);
         }
         return routingKey;
     }
 
     /**
      * Resolve the Routing Key for the given {@code command}.
      *
-     * @param command The command to resolve the routing key for
-     * @return the String representing the Routing Key, or {@code null} if unresolved.
+     * @param command the command to resolve the routing key for
+     * @return the String representing the Routing Key, or {@code null} if unresolved
      */
     protected abstract String doResolveRoutingKey(CommandMessage<?> command);
+
```

Comment:
Since this is an abstract class, I doubt whether using the builder pattern here simplifies things.

Reply:
Not necessarily simplify, no, but giving us the flexibility to change whatever without creating [5/10/N] new constructors. So from the same stance as to why the other infra components have a builder instead of a constructor. Let's chat offline whether we should go this route for components of this size.